### PR TITLE
New resource caching system

### PR DESCRIPTION
**Pull request by psav**
Implement new resource caching provider system

* Added new GetPrimaryLabel() to ClowdObject types to help with logging etc
* Added new providers for confighash, cronjob, dependencies, deployment,
  metrics, serviceaccount, servicemesh and web.
* Removed maker system completely.
* Removed proxyClient.
* Rejigged app/env reconciliation functions.
* Added new resource cache system.
* Added resourceIdents for all touched resources.

### CODE REVIEW

File: controllers/cloud.redhat.com/providers/providers.go [hunk 1]
```diff
@@ -134,3 +211,236 @@ func ExtractSecretData(secrets []core.Secret, fn ExtractFn, keys ...string) {
...
+// before modifying the obejct they wish to be placed in the cache.
+func (o *ObjectCache) Create(providerIdent ProviderIdent, nn types.NamespacedName, object runtime.Object) error {
+
+	update, err := utils.UpdateOrErr(o.client.Get(o.ctx, nn, object))
+
+	if err != nil {
+		return err
+	}
+
+	if _, ok := o.data[providerIdent][nn]; ok {
```
Comment (psav) [inline controllers/cloud.redhat.com/providers/providers.go:272]: test this !

File: controllers/cloud.redhat.com/providers/resource_cache.go [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+func (o *ObjectCache) ApplyAll() error {
+	for k, v := range o.data {
+		for n, i := range v {
+			err := i.Update.Apply(o.ctx, o.client, i.Object)
+			o.log.Info("APPLY resource ", "namespace", n.Namespace, "name", n.Name, "provider", k.Provider, "purpose", k.Purpose, "kind", i.Object.GetObjectKind().GroupVersionKind().Kind, "update", i.Update)
+			if err != nil {
+				return err
+			}
+			if i.Status {
+				o.client.Status().Update(o.ctx, i.Object)
```
Comment (kylape) [inline controllers/cloud.redhat.com/providers/resource_cache.go:259]: this this return an error?

**Review by kylape — CHANGES_REQUESTED** [review]
Just reviewed controllers and resource cache

File: controllers/cloud.redhat.com/clowdapp_controller.go [hunk 3]
```diff
@@ -201,7 +207,16 @@ func ignoreStatusUpdatePredicate() predicate.Predicate {
...
-					if objOld.Status.Ready == false && objNew.Status.Ready == true {
+					if !objOld.Status.Ready && objNew.Status.Ready {
+						return true
+					}
+				}
+			}
+
+			if objOld, ok := e.ObjectOld.(*strimzi.Kafka); ok {
+				if objNew, ok := e.ObjectNew.(*strimzi.Kafka); ok {
+					// Can cause a panic
```
[the remark this answers is not in the document]
Comment (psav) [reply controllers/cloud.redhat.com/clowdapp_controller.go:228]: Resolved with the line below - I updated it a while back

File: controllers/cloud.redhat.com/clowdenvironment_controller.go [hunk 4]
```diff
@@ -121,21 +124,28 @@ func (r *ClowdEnvironmentReconciler) Reconcile(req ctrl.Request) (ctrl.Result, e
...
+		Client: r.Client,
 		Env:    &env,
+		Cache:  &cache,
 	}
 
+	var requeue bool
+
 	err = runProvidersForEnv(provider)
+	if err != nil {
+		requeue = errors.HandleError(ctx, err)
```
[the remark this answers is not in the document]
Comment (psav) [reply controllers/cloud.redhat.com/clowdenvironment_controller.go]: Not any more sir

File: controllers/cloud.redhat.com/providers/resource_cache.go [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+						err := o.client.Delete(o.ctx, &obj)
+						if err != nil {
+							return err
+						}
+					}
+				}
+			}
+		}
+	}
+	//fmt.Println("\n-----------------")
```
[the remark this answers is not in the document]
Comment (psav) [reply controllers/cloud.redhat.com/providers/resource_cache.go:401]: Yes these have been helpful in debugging so until we were ready to merge, I was keeping them around

File: controllers/cloud.redhat.com/providers/metrics/utils.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import (
+	crd "cloud.redhat.com/clowder/v2/apis/cloud.redhat.com/v1alpha1"
+	deployProvider "cloud.redhat.com/clowder/v2/controllers/cloud.redhat.com/providers/deployment"
+	webProvider "cloud.redhat.com/clowder/v2/controllers/cloud.redhat.com/providers/web"
+	apps "k8s.io/api/apps/v1"
+	core "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+func (m *metricsProvider) makeMetrics(deployment *crd.Deployment, app *crd.ClowdApp) error {
```
Comment (kylape) [inline controllers/cloud.redhat.com/providers/metrics/utils.go:12]: does this need to be in a separate file from the Provide fn?  It doesn't seem to be used anywhere else.

File: controllers/cloud.redhat.com/providers/serviceaccount/utils.go [hunk 7]
```diff
@@ -1,28 +1,29 @@
...
 	"cloud.redhat.com/clowder/v2/controllers/cloud.redhat.com/object"
+	"cloud.redhat.com/clowder/v2/controllers/cloud.redhat.com/providers"
 	"cloud.redhat.com/clowder/v2/controllers/cloud.redhat.com/utils"
 	core "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/types"
-	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-func createServiceAccount(ctx context.Context, client client.Client, obj object.ClowdObject, pullSecretNames crd.PullSecrets) error {
+func createServiceAccount(cache *providers.ObjectCache, ident providers.ResourceIdent, obj object.ClowdObject, pullSecretNames crd.PullSecrets) error {
```
Comment (kylape) [inline controllers/cloud.redhat.com/providers/serviceaccount/utils.go:13]: does this need to be in a separate utils file?

File: controllers/cloud.redhat.com/providers/deployment/utils.go [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+			MaxUnavailable: &intstr.IntOrString{Type: intstr.Int, IntVal: int32(25)},
+		},
+	}
+	d.Spec.ProgressDeadlineSeconds = utils.Int32Ptr(600)
+
+	d.Spec.Template.Spec.ImagePullSecrets = []core.LocalObjectReference{
+		{Name: "quay-cloudservices-pull"},
+	}
+
+	d.Spec.Template.Spec.ServiceAccountName = app.GetClowdSAName()
```
Comment (kylape) [inline controllers/cloud.redhat.com/providers/deployment/utils.go:64]: should this be set in the service account provider?

File: controllers/cloud.redhat.com/utils/utils.go [hunk 9]
```diff
@@ -59,15 +64,15 @@ func (u *Updater) Apply(ctx context.Context, cl client.Client, obj runtime.Objec
 	meta := obj.(metav1.Object)
 
 	if *u {
-		Log.Info("Updating resource", "namespace", meta.GetNamespace(), "name", meta.GetName(), "kind", kind)
+		//Log.Info("Updating resource", "namespace", meta.GetNamespace(), "name", meta.GetName(), "kind", kind)
```
Comment (kylape) [inline controllers/cloud.redhat.com/utils/utils.go:67]: should this line be removed?

File: controllers/cloud.redhat.com/utils/utils.go [hunk 10]
```diff
@@ -59,15 +64,15 @@ func (u *Updater) Apply(ctx context.Context, cl client.Client, obj runtime.Objec
...
 		err = cl.Update(ctx, obj)
 	} else {
 		if meta.GetName() == "" {
 			Log.Info("Skipping resource as name unknown", "kind", kind)
 			return nil
 		}
 
-		Log.Info("Creating resource", "namespace", meta.GetNamespace(), "name", meta.GetName(), "kind", kind)
+		//Log.Info("Creating resource", "namespace", meta.GetNamespace(), "name", meta.GetName(), "kind", kind)
```
Comment (kylape) [inline controllers/cloud.redhat.com/utils/utils.go:75]: should this line be removed?

**Review by kylape — CHANGES_REQUESTED** [review]
what does the always.go naming convention supposed to represent?

lots of `utils.go` files that seem like they could be combined with `always.go`.  It's not an issue per se, but I'm curious what the thinking is.

File: controllers/cloud.redhat.com/providers/deployment/utils.go (hunk 8 above)
Comment (psav) [reply controllers/cloud.redhat.com/providers/deployment/utils.go:64]: Yes - this is happening in other places too - we should certainly deal with this......probably in this PR, but we need to sort out how to pull these deployments out

File: controllers/cloud.redhat.com/providers/metrics/utils.go (hunk 6 above)
Comment (psav) [reply controllers/cloud.redhat.com/providers/metrics/utils.go:12]: it was for consistency and future expansion

File: controllers/cloud.redhat.com/providers/serviceaccount/utils.go (hunk 7 above)
Comment (psav) [reply controllers/cloud.redhat.com/providers/serviceaccount/utils.go:13]: no - once again consistency

File: controllers/cloud.redhat.com/utils/utils.go (hunk 9 above)
Comment (psav) [reply controllers/cloud.redhat.com/utils/utils.go:67]: I wasn't sure if we should or not - we log earlier - this is a very low level log

File: controllers/cloud.redhat.com/utils/utils.go (hunk 10 above)
Comment (psav) [reply controllers/cloud.redhat.com/utils/utils.go:75]: I wasn't sure if we should or not - we log earlier - this is a very low level log

**Review by kylape — APPROVED** [review]

**Pull request merged**